contracts/mixnet/src/interval/pending_events.rs

```diff
@@ -0,0 +1,403 @@
+// Copyright 2022 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::delegations;
+use crate::delegations::storage as delegations_storage;
+use crate::interval::helpers::change_epochs_in_interval;
+use crate::interval::storage;
+use crate::mixnet_contract_settings::storage as mixnet_params_storage;
+use crate::mixnodes::helpers::{cleanup_post_unbond_mixnode_storage, get_mixnode_details_by_id};
+use crate::rewards::storage as rewards_storage;
+use crate::support::helpers::send_to_proxy_or_owner;
+use cosmwasm_std::{wasm_execute, Addr, Coin, DepsMut, Env, Response};
+use mixnet_contract_common::error::MixnetContractError;
+use mixnet_contract_common::events::{
+    new_active_set_update_event, new_delegation_event, new_delegation_on_unbonded_node_event,
+    new_interval_config_update_event, new_mixnode_cost_params_update_event,
+    new_mixnode_unbonding_event, new_rewarding_params_update_event, new_undelegation_event,
+};
+use mixnet_contract_common::mixnode::MixNodeCostParams;
+use mixnet_contract_common::pending_events::{PendingEpochEvent, PendingIntervalEvent};
+use mixnet_contract_common::reward_params::IntervalRewardingParamsUpdate;
+use mixnet_contract_common::rewarding::helpers::truncate_reward_amount;
+use mixnet_contract_common::{Delegation, NodeId};
+use std::time::Duration;
+use vesting_contract_common::messages::ExecuteMsg as VestingContractExecuteMsg;
+
+pub(crate) trait ContractExecutableEvent {
+    // note: the error only means a HARD error like we failed to read from storage.
+    // if, for example, delegating fails because mixnode no longer exists, we return an Ok(()),
+    // because it's not a hard error and we don't want to fail the entire transaction
+    fn execute(self, deps: DepsMut<'_>, env: &Env) -> Result<Response, MixnetContractError>;
+}
+
+pub(crate) fn delegate(
+    deps: DepsMut<'_>,
+    env: &Env,
+    owner: Addr,
+    mix_id: NodeId,
+    mut amount: Coin,
+    proxy: Option<Addr>,
+) -> Result<Response, MixnetContractError> {
+    // check if the target node still exists (it might have unbonded between this event getting created
+    // and being executed). Do note that it's absolutely possible for a mixnode to get immediately
+    // unbonded at this very block (if the event was pending), but that's tough luck, then it's up
+    // to the delegator to click the undelegate button
+    let mixnode_details = match get_mixnode_details_by_id(deps.storage, mix_id)? {
+        Some(details)
+            if details.rewarding_details.still_bonded()
+                && !details.bond_information.is_unbonding =>
+        {
+            details
+        }
+        _ => {
+            // if mixnode is no longer bonded or in the process of unbonding, return the tokens back to the
+            // delegator;
+            // TODO: @DU: do we need to do any vesting-specific tracking here?
```

Comment:
Yes, probably easiest is to send a `TrackUndelegate` message, as if it was a regular delegation